### ISSUE TITLE
Composite device support

### ISSUE THREAD

**Issue by sjroe** [rust-embedded-community/usbd-serial#3]
For this to work as part of a composite device an [interface association descriptor](https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-interface-association-descriptor) (IAD) is required which in turn requires that the device class, sub class and function are set as 0xEF, 0x02 and 0x01 respectively.

I achieved producing a composite device with two CDC ACM serial ports by:

- Setting the device class etc as above
- Writing the IAD directly before the interface descriptors
- Enabling feature `control-buffer-256`

The IAD was initially inserted directly into `get_configuration_descriptors` as per the below:

```rust
writer.write(
    INTERFACE_ASSOCIATION, // 0x0B
    &[
        self.com_if.into(), // bFirstInterface
        2, // bInterfaceCount
        USB_CLASS_CDC, // bfunctionClass
        CDC_SUBCLASS_ACM, // bfunctionSubClass
        CDC_PROTOCOL_NONE, // bfunctionProtocol
        0, // iFunction
    ])?;
```

I have since refactored this a little extending usb-device to include an `interface_association()` function much like `interface()` and `endpoint()` as this is useful across multiple classes of device:

```rust
pub fn interface_association(&mut self, first_interface: InterfaceNumber,
    interface_count: u8, function_class: u8, function_sub_class: u8, function_protocol: u8)
    -> Result<()>
{
    self.write(
        descriptor_type::INTERFACE_ASSOCIATION,
        &[
            first_interface.into(), // bFirstInterface
            interface_count, // bInterfaceCount
            function_class, // bfunctionClass
            function_sub_class, // bfunctionSubClass
            function_protocol, // bfunctionProtocol
            0, // iFunction
        ])?;

    Ok(())
}
```

This IAD is only required for composite devices.  What do you think the best approach would be to only include the IAD for usbd-serial when needed?

**Comment by mvirkkunen** [comment]
Hmm. Interesting. I wonder if having the IAD there always would cause any problems, even if it's not a composite device? It's not that many bytes... I guess it's worth investigating if it causes any compatibility issues.

**Comment by sjroe** [comment]
So I just tested having the IAD present for a non-composite device on Windows 10 and it successfully enumerated without any issue.

However, usbview.exe displays an error in its parsed view:

```
          ===>Device Descriptor<===
bLength:                           0x12
bDescriptorType:                   0x01
bcdUSB:                          0x0200
bDeviceClass:                      0x02
*!*ERROR: device class should be Multi-interface Function 0xEF
          When IAD descriptor is used
bDeviceSubClass:                   0x00
*!*ERROR: device SubClass should be USB Common Sub Class 2
          When IAD descriptor is used
bDeviceProtocol:                   0x00
*!*ERROR: device Protocol should be USB IAD Protocol 1
          When IAD descriptor is used
```

I think that including the IAD on a non-composite device is non-standard and may have issues on other OSs/OS versions.  If you want to go this route then exhaustive testing would be required.

Ideally the classes would have an awareness of the device configuration allowing them to selectively include IADs (and other things?) when required.

**Comment by mvirkkunen** [comment]
Hmm. Maybe in light of all this it might make sense to move the class/subclass/protocol values completely into the classes as opposed to having the user set them in the device configuration. At configuration time `usb-device` itself could check the number of classes to see if the device is a composite device and then the classes could co-operate to write out the IADs if needed. This is of course assuming that classes behave well enough so that this automation is feasible...

**Comment by mvirkkunen** [comment]
This was implemented in #13.
